packages/donate-button-v4/src/autoPlayMode.tsx

```diff
@@ -0,0 +1,178 @@
+import {WidgetLoader} from 'src/loaders/Widgetloader';
+import resetcss from 'src/resetCss';
+
+const OBSERVER_OPTIONS = {
+	childList: true,
+	subtree: true
+};
+
+interface CreateButtonProps extends Partial<WidgetConfig> {
+	element: Element;
+	nonprofitSlug: string;
+	onClick: () => void;
+}
+
+interface CreateWidgetProps extends Partial<WidgetConfig> {
+	nonprofitSlug: string;
+}
+
+let shadowRoot: ShadowRoot;
+let widgetContainer: HTMLElement;
+
+class WidgetController {
+	options: CreateWidgetProps;
+	mountPoint: HTMLElement;
+	constructor(options: CreateWidgetProps, mountPoint: HTMLElement) {
+		this.options = options;
+		this.mountPoint = mountPoint;
+	}
+
+	hide() {
+		this.options.show = false;
+		this.render();
+	}
+
+	show() {
+		this.options.show = true;
+		this.render();
+	}
+
+	setOptions(options: WidgetConfig) {
+		this.options = options;
+		this.render();
+	}
+
+	render() {
+		preactRender(
+			<WidgetLoader
+				options={this.options}
+				hide={() => {
+					this.hide();
+				}}
+			/>,
+			this.mountPoint
+		);
+	}
+}
+
+const parseUrl = (
+	urlString: string
+): (Partial<WidgetConfig> & {nonprofitSlug: string}) | undefined => {
+	const url = new URL(urlString);
+
+	const fundraiserSlug = url.pathname.split('/f/')[1];
+	const nonprofitSlug = url.pathname.split('/')[1];
+
+	if (!nonprofitSlug) {
+		return;
+	}
+
+	return {
+		fundraiserSlug,
+		nonprofitSlug
+	};
+};
+
+function mountShadowRoot() {
+	const shadowWidgetWrapper = document.createElement('div');
+	shadowWidgetWrapper.id = 'shadow-wrapper';
+	document.body.append(shadowWidgetWrapper);
+	shadowRoot = shadowWidgetWrapper.attachShadow({mode: 'open'});
+}
+
+function mountWidgetContainer() {
+	if (!shadowRoot) {
+		mountShadowRoot();
+	}
+
+	widgetContainer = document.createElement('div');
+	widgetContainer.id = 'widget-container';
+	shadowRoot.append(widgetContainer);
+
+	const edoStyles: HTMLStyleElement | null =
+		document.querySelector('#edo-styles');
+
+	if (edoStyles) {
+		const rules = Object.values(edoStyles.sheet?.cssRules ?? {})
+			.map((rule) => rule.cssText)
+			.join('\n');
+
+		const edoShadowStyles = document.createElement('style');
+		edoShadowStyles.id = 'edo-shadow-styles';
+		edoShadowStyles.innerHTML = resetcss + rules;
+
+		widgetContainer.append(edoShadowStyles);
+	}
+}
+
+function createWidget(options: CreateWidgetProps) {
+	if (!widgetContainer) {
+		mountWidgetContainer();
+	}
+
+	const widgetMountPoint = document.createElement('div');
+	widgetContainer.append(widgetMountPoint);
+
+	return new WidgetController(options, widgetMountPoint);
+}
+
+function createButton({element, ...options}: CreateButtonProps) {
+	if (!element.parentNode) {
+		return;
+	}
+
+	const buttonContainer = element.parentNode as HTMLElement;
+
+	preactRender(<EmbedButton {...options} />, buttonContainer, element);
+}
+
+export default function autoPlayMode() {
+	loadFonts();
+
+	const observer = new MutationObserver((_, observer) => {
+		observer.disconnect();
```

Comment:
Why do you `disconnect` at the top, then `observe` again at the end? What if some relevant changes come in between those two calls?

If good idea to keep, please document with an in-code comment.

Reply:
because we change DOM inside observer callback. so we need disconnect to not cause infinite loop